drivers/network/netio/netio.c

```diff
@@ -0,0 +1,896 @@
+    if (TargetConnectionInfo == NULL)
+    {
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    /* TODO: which address?? */
+    Ignored = TdiConnectionInfoFromSocketAddress(RemoteAddress);
+    if (Ignored == NULL)
+    {
+        ExFreePoolWithTag(TargetConnectionInfo, TAG_AFD_TDI_CONNECTION_INFORMATION);
+
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    /* Some drivers (like WinDRBD) do not set this.
+     * But ReactOS seems to require this (else BSOD on completion).
+     */
+
+    if (Irp->Tail.Overlay.Thread == NULL)
+    {
+        Irp->Tail.Overlay.Thread = PsGetCurrentThread();
+    }
+    IoMarkIrpPending(Irp);
+    SocketGet(s);
+    InsertTailList(&s->PendingUserIrps, &Irp->Tail.Overlay.ListEntry);
+
+    status = TdiConnect(&tdiIrp, s->ConnectionFile, TargetConnectionInfo, Ignored, NetioComplete, nc);
+    /* TODO: check this again..I would assume if the TDI helper function
+       returns any error, the completion function is not called.
+     */
+    if (!NT_SUCCESS(status))
+    {
+        SocketPut(s);
+        ExFreePoolWithTag(nc, 'NEIO');
+    }
+    return status;
+}
+
+static NTSTATUS WSKAPI
+WskStreamIo(_In_ PWSK_SOCKET Socket,
+            _In_ PWSK_BUF Buffer, _In_ ULONG Flags, _Inout_ PIRP Irp, enum direction Direction)
+{
+    PIRP tdiIrp = NULL;
+    struct _WSK_SOCKET_INTERNAL *s = (struct _WSK_SOCKET_INTERNAL *)Socket;
+    NTSTATUS status;
+    void *BufferData;
+    struct NetioContext *nc;
+
+    /* Call ourselves. Sets status to pending. The interesting
+     * part happens later.
+     */
+    DummyCallDriver(Irp);
+
+    nc = ExAllocatePoolWithTag(NonPagedPool, sizeof(*nc), 'NEIO');
+    if (nc == NULL)
+    {
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+    nc->socket = s;
+    nc->UserIrp = Irp;
+
+    BufferData = MmGetSystemAddressForMdlSafe(Buffer->Mdl, NormalPagePriority);
+    if (BufferData == NULL)
+    {
+        DbgPrint("Error mapping MDL\n");
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    /* Some drivers (like WinDRBD) do not set this.
+     * But ReactOS seems to require this (else BSOD on completion).
+     */
+
+    if (Irp->Tail.Overlay.Thread == NULL)
+    {
+        Irp->Tail.Overlay.Thread = PsGetCurrentThread();
+    }
+    IoMarkIrpPending(Irp);
+    /* TODO: protect by spinlock ... */
+    InsertTailList(&s->PendingUserIrps, &Irp->Tail.Overlay.ListEntry);
+    SocketGet(s);
+
+    if (Direction == DIR_SEND)
+    {
+        /* This will create a tdiIrp: */
+        status =
+            TdiSend(&tdiIrp, s->ConnectionFile, 0, ((char *)BufferData) + Buffer->Offset, Buffer->Length,
+                    NetioComplete, nc);
```

Comment:
```suggestion
        status = TdiSend(&tdiIrp, s->ConnectionFile, 0, ((char *)BufferData) + Buffer->Offset,
                         Buffer->Length, NetioComplete, nc);
```
shorter lines and one line less.